### PR TITLE
Transforms: Labels to fields rewrite that uses merge transform inside it 

### PR DESCRIPTION
**Pull request by torkelo**
Looking at solving a graphite labels grouping issue  (https://github.com/grafana/grafana/issues/15279)  and wanted to try to add an option to labels to fields transform. 

Discovered a big possible simplification to the labels to fields transform. If all this transform did was operate on single data frames and extract the labels to new fields, and then use the merge transform (that did not exist when we wrote labels to fields). Then labels to fields can be made much simpler. 

thoughts @hugohaggmark @mckn ?

### CODE REVIEW

**Comment by hugohaggmark** [comment]
Taking it for a spin now

File: packages/grafana-data/src/transformations/transformers/labelsToFields.test.ts [hunk 1]
```diff
@@ -1,232 +1,104 @@
...
-      { name: 'feelsLike', type: FieldType.string, values: new ArrayVector(['ok', null, 'ok', null]), config: {} },
-      { name: 'sky', type: FieldType.string, values: new ArrayVector([null, 'cloudy', null, 'cloudy']), config: {} },
-      { name: 'temp', type: FieldType.number, values: new ArrayVector([1, -1, 2, -2]), config: {} },
+    const expected: FieldDTO[] = [
+      { name: 'time', type: FieldType.time, values: [1000, 2000], config: {} },
+      { name: 'location', type: FieldType.string, values: ['inside', 'outside'], config: {} },
+      { name: 'temp', type: FieldType.number, values: [1, -1], config: {} },
     ];
 
-    expect(result[0].fields).toEqual(expected);
```
Comment (hugohaggmark) [inline packages/grafana-data/src/transformations/transformers/labelsToFields.test.ts:230]: Are these tests covered by the `merge` transformer? Those tests tested data frames with different labels.

File: packages/grafana-data/src/transformations/transformers/labelsToFields.ts [hunk 2]
```diff
@@ -1,173 +1,61 @@
...
+    for (const frame of data) {
+      const newFields: Field[] = [];
 
-    const timeColumnItem = columnsMap[fields[0].name];
-    const seriesIndexStrings = Object.keys(timeColumnItem);
-    for (const seriesIndexString of seriesIndexStrings) {
-      const seriesItem = timeColumnItem[seriesIndexString];
-      const timeValueStrings = Object.keys(seriesItem.values);
+      for (const field of frame.fields) {
+        if (field.labels) {
```
Comment (hugohaggmark) [inline packages/grafana-data/src/transformations/transformers/labelsToFields.ts:25]: nitpick: maybe reverse `if` statement and continue to minimize nesting?

File: packages/grafana-data/src/transformations/transformers/labelsToFields.ts [hunk 3]
```diff
@@ -1,173 +1,61 @@
...
+              config: {},
+            });
+          }
 
-      for (const timeValueString of timeValueStrings) {
-        if (!values[timeValueString]) {
-          values[timeValueString] = [];
+          const newField = { ...field, name };
+          delete newField.labels;
+          delete newField.config.displayName;
```
Comment (hugohaggmark) [inline packages/grafana-data/src/transformations/transformers/labelsToFields.ts:46]: This looks weird, maybe add a comment.

**Review by hugohaggmark — COMMENTED** [review]
Looks good ⭐ , love that you could reduce the code so much. But some tests were removed that need some explanation

File: packages/grafana-data/src/transformations/transformers/labelsToFields.test.ts (hunk 1 above)
Comment (torkelo) [reply packages/grafana-data/src/transformations/transformers/labelsToFields.test.ts:230]: yes, it has tests with data frames with different fields

File: packages/grafana-data/src/transformations/transformers/labelsToFields.ts (hunk 3 above)
Comment (torkelo) [reply packages/grafana-data/src/transformations/transformers/labelsToFields.ts:46]: yea, do no love this either. But want to remove these on the value field after the transform. But setting them to undefined is also ugly as it leaves the props there in the object. Maybe that is ok

File: public/app/core/components/TransformersUI/LabelsToFieldsTransformerEditor.tsx [hunk 4]
```diff
@@ -8,14 +15,52 @@ export const LabelsAsFieldsTransformerEditor: React.FC<TransformerUIProps<Labels
...
+        <Select
+          isClearable={true}
+          allowCustomValue={false}
+          placeholder="(Optional) Select label"
+          options={labelNames}
+          className="min-width-18 gf-form-spacing"
+          value={options?.valueLabel}
+          onChange={onValueLabelChange}
+          menuPlacement="bottom"
+        />
```
Comment (dprokop) [inline public/app/core/components/TransformersUI/LabelsToFieldsTransformerEditor.tsx:53]: Would be great to add some information about value field name behavior in the UI

**Review by dprokop — COMMENTED** [review]
Code-wise looks good, great that you have removed so much complexity there! Think docs need some update as well after introducing new option: https://grafana.com/docs/grafana/latest/panels/transformations/#labels-to-fields

File: packages/grafana-data/src/transformations/transformers/labelsToFields.ts (hunk 2 above)
Comment (torkelo) [reply packages/grafana-data/src/transformations/transformers/labelsToFields.ts]: fixed

**Comment by torkelo** [comment]
Dam merged the wrong PR, need to create a follow up PR with docs and refining this

**Pull request merged**